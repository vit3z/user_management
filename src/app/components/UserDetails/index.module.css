.userDetailsModalOverlay {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(0, 0, 0, 0.5);
  display: flex;
  justify-content: center;
  align-items: center;
  z-index: 1000;
}

.userDetailsModalContent {
  display: flex;
  flex-direction: row;
  margin: 16px;
  background: white;
  border-radius: 8px;
  box-shadow: 0 2px 10px rgba(0, 0, 0, 0.1);
  position: relative;
  width: 750px;
}

.userDetailsModalHeaderSection {
  align-content: flex-start;
  padding: 32px 16px 0px;
  border-top-left-radius: 8px;
  border-bottom-left-radius: 8px;
  background-color: #FFA817;
}

.userDetailsModalUserAvatar {
  display: flex;
  justify-content: center;
  margin: auto;
  border-radius: 50%;
}

.userDetailsModalContentUserName {
  color: #46596B;
  font-weight: 300;
  display: flex;
  justify-content: center;
  text-align: center;
}

.userDetailsModalContentUserWebsite {
  color: #46596B;
  font-weight: 300;
  display: flex;
  justify-content: center;
  padding-bottom: 8px;
}

.userDetailsModalDivider {
  display: none;
}

.userDetailsModalBodySection {
  padding: 16px;
  width: 70%;
}

.userDetailsModalContentSectionTitle {
  color: #46596B;
  font-weight: 300;
  margin-top: 16px;
  text-decoration: underline;
}

.userDetailsModalContentSectionInfo {
  color: #4b5b69;
  font-weight: 400;
  font-size: 16px;
}

.userDetailsModalContentSectionInfoQuote {
  color: #4b5b69;
  font-style: italic;
}

.userDetailsModalContentSectionCentered {
  color: #4b5b69;
  font-weight: 400;
  display: flex;
  justify-content: center;
}

.userDetailsModalContentSectionInfoSmallText {
  color: #606366;
  font-weight: 300;
  font-size: 14px;
}

.userDetailsModalCloseButton {
  margin-top: 16px;
  margin-bottom: 16px;
  padding: 8px 16px;
  background-color: #178BFF;
  color: white;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  float: right;
}

.userDetailsModalCloseButton:hover {
  background-color: #1742FF;
  -webkit-transition: background-color 250ms ease-in;
  -ms-transition: background-color 250ms ease-in;
  transition: background-color 250ms ease-in;
}

@media (max-width: 1030px) {
  .userDetailsModalContent {
    width: 80%;
    display: flex;
    flex-direction: column;
  }

  .userDetailsModalHeaderSection {
    padding-top: 16px;
    border-top-left-radius: 8px;
    border-top-right-radius: 8px;
    border-bottom-left-radius: 0px;
    background-color: #FFA817;
  }

  .userDetailsModalDivider {
    display: inherit;
    border-top: 1px solid #cecfdb;
  }

  .userDetailsModalBodySection {
    width: 100%;
  }

  .userDetailsModalCloseButton {
    width: 100%;
  }
}