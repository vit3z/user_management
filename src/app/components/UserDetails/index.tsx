import React, { ReactNode, useEffect } from "react";
import styles from "./index.module.css";
import { UserInformationProps } from "@/consts/interfaces";
import Image from "next/image";
import ProfileImage from "../../../../public/Profile_avatar_placeholder_large.png";

interface UserDetailsProps {
  isOpen: boolean;
  onClose: () => void;
  data: UserInformationProps;
}

const UserDetails: React.FC<UserDetailsProps> = ({ isOpen, onClose, data }) => {
  useEffect(() => {
    const handleEsc = (event: KeyboardEvent) => {
      if (event.key === "Escape") {
        onClose();
      }
    };
    document.addEventListener("keydown", handleEsc);
    return () => {
      document.removeEventListener("keydown", handleEsc);
    };
  }, [onClose]);

  if (!isOpen) {
    return null;
  }

  return (
    <div className={styles.userDetailsModalOverlay} onClick={onClose}>
      <div
        className={styles.userDetailsModalContent}
        onClick={(e) => e.stopPropagation()}
      >
        <div className={styles.userDetailsModalHeaderSection}>
          <Image
            className={styles.userDetailsModalUserAvatar}
            src={ProfileImage}
            width={50}
            height={50}
            alt="User Avatar"
          />
          <h2 className={styles.userDetailsModalContentUserName}>
            {data.name}
          </h2>
          <a
            href={`https://google.com `}
            target="_blank"
            className={styles.userDetailsModalContentUserWebsite}
          >
            {data.website}
          </a>
          <hr className={styles.userDetailsModalDivider} />
        </div>

        <div className={styles.userDetailsModalBodySection}>
          <h3 className={styles.userDetailsModalContentSectionTitle}>
            Contact Details
          </h3>
          <p className={styles.userDetailsModalContentSectionInfo}>
            <a href={`mailto:${data.email}`}>
              Email: <b>{data.email}</b>{" "}
            </a>
          </p>
          <p className={styles.userDetailsModalContentSectionInfo}>
            <a href={`tel:${data.phone}`}>
              Phone: <b>{data.phone}</b>
            </a>
          </p>
          <p className={styles.userDetailsModalContentSectionInfo}>
            Username: <b>{data.username}</b>
          </p>

          <h3 className={styles.userDetailsModalContentSectionTitle}>
            Company
          </h3>
          <p className={styles.userDetailsModalContentSectionInfo}>
            <b>{data.company.name}</b>
          </p>
          <p className={styles.userDetailsModalContentSectionInfo}>
            Summary: <b>{data.company.bs}</b>
          </p>
          <p className={styles.userDetailsModalContentSectionInfo}>
            <b>{data.company.catchPhrase}</b>
          </p>

          <h3 className={styles.userDetailsModalContentSectionTitle}>
            Address
          </h3>
          <p className={styles.userDetailsModalContentSectionInfo}>
            <b>{`${data.address.suite} ${data.address.street}, ${data.address.city}`}</b>
          </p>
          <p className={styles.userDetailsModalContentSectionInfo}>
            <b>{data.address.zipcode}</b>
          </p>
          <p
            className={styles.userDetailsModalContentSectionInfoSmallText}
          >{`${data.address.geo.lat}, ${data.address.geo.lng}`}</p>
          <button
            className={styles.userDetailsModalCloseButton}
            onClick={onClose}
          >
            Close
          </button>
        </div>
      </div>
    </div>
  );
};

export default UserDetails;
