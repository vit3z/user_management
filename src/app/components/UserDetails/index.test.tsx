import '@testing-library/jest-dom'
import { render, fireEvent, cleanup } from '@testing-library/react'
import UserDetails from '.'

describe("UserDetails", () => {
  const basicUserDetails = {
    id: 125,
    name: "John",
    username: "Doe",
    email: "jd@jd.jd",
    address: {
      street: "Test Avenue",
      suite: "11",
      city: "London",
      zipcode: "EC1",
      geo: {
        lat: "test lat",
        lng: "test lng",
      }
    },
    phone: "555-5555",
    website: "test.com",
    company: {
      name: "Company",
      catchPhrase: "Phrase",
      bs: "We do it",
    }
  }

  const onClose = jest.fn()

  describe("renders all elements when needed", () => {
    it("renders the component when props are passed", () => {
      const {container} = render(<UserDetails isOpen={true} onClose={onClose} data={basicUserDetails}/>);
      expect(container).toBeInTheDocument();
      expect(container).toMatchSnapshot();
    })
  
    it("does not render the user ID when props are passed", () => {
      const {queryByText} = render(<UserDetails isOpen={true} onClose={onClose} data={basicUserDetails}/>);
      const userID = queryByText('125');
      expect(userID).not.toBeInTheDocument();
    })
  
    it("renders the user name when props are passed", () => {
      const {queryByText} = render(<UserDetails isOpen={true} onClose={onClose} data={basicUserDetails}/>);
      const userFullName = queryByText('John');
      expect(userFullName).toBeInTheDocument();
    })
  
    it("renders the user website when props are passed", () => {
      const {getByText} = render(<UserDetails isOpen={true} onClose={onClose} data={basicUserDetails}/>);
      const userWebsite = getByText('test.com');
      expect(userWebsite).toBeInTheDocument();
    })
  
    it("renders the user contact details when props are passed", () => {
      const {getByText} = render(<UserDetails isOpen={true} onClose={onClose} data={basicUserDetails}/>);
      const userEmail = getByText('jd@jd.jd');
      expect(userEmail).toBeInTheDocument();
  
      const userPhone = getByText('555-5555');
      expect(userPhone).toBeInTheDocument();
  
      const userUsername = getByText('Doe');
      expect(userUsername).toBeInTheDocument();
    })
  
    it("renders the user company details when props are passed", () => {
      const {getByText} = render(<UserDetails isOpen={true} onClose={onClose} data={basicUserDetails}/>);
      const userCompany = getByText(/Company/i);
      expect(userCompany).toBeInTheDocument();
  
      const userSummary = getByText(/We do it/i);
      expect(userSummary).toBeInTheDocument();
    })
  
    it("renders the user address when props are passed", () => {
      const {getByText} = render(<UserDetails isOpen={true} onClose={onClose} data={basicUserDetails}/>);
      const userAddress = getByText('11 Test Avenue, London');
      expect(userAddress).toBeInTheDocument();
  
      const userZipCode = getByText('EC1');
      expect(userZipCode).toBeInTheDocument();
      
      const userGeoLoc = getByText('test lat, test lng');
      expect(userGeoLoc).toBeInTheDocument();
    })
  
    it("does not render user details if isOpen is false", () => {
      const {queryByText} = render(<UserDetails isOpen={false} onClose={onClose} data={basicUserDetails}/>);
      const userFullName = queryByText('John');
      expect(userFullName).not.toBeInTheDocument();
    })
  })

  describe("triggers functions when needed", () => {
    afterEach(cleanup);

    it('calls onClose when Escape key is pressed', () => {
      const onCloseMock = jest.fn();
      render(<UserDetails isOpen={true} onClose={onCloseMock} data={basicUserDetails}/>);
  
      // Simulate pressing the Escape key
      fireEvent.keyDown(document, { key: 'Escape', code: 'Escape', keyCode: 27, charCode: 27 });
  
      expect(onCloseMock).toHaveBeenCalledTimes(1);
    });

    it('removes the event listener on unmount', () => {
      const onCloseMock = jest.fn();
      const {unmount} = render(<UserDetails isOpen={true} onClose={onCloseMock} data={basicUserDetails}/>);
  
      // Unmount the component
      unmount();
  
      // Simulate pressing the Escape key after unmount
      fireEvent.keyDown(document, { key: 'Escape', code: 'Escape', keyCode: 27, charCode: 27 });
  
      expect(onCloseMock).not.toHaveBeenCalled();
    });
  })
})