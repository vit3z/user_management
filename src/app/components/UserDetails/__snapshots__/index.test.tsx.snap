// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDetails renders all elements when needed renders the component when props are passed 1`] = `
<div>
  <div
    class="userDetailsModalOverlay"
  >
    <div
      class="userDetailsModalContent"
    >
      <div
        class="userDetailsModalHeaderSection"
      >
        <img
          alt="User Avatar"
          class="userDetailsModalUserAvatar"
          data-nimg="1"
          decoding="async"
          height="50"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=128&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=64&q=75 1x, /_next/image?url=%2Fimg.jpg&w=128&q=75 2x"
          style="color: transparent;"
          width="50"
        />
        <h2
          class="userDetailsModalContentUserName"
        >
          John
        </h2>
        <a
          class="userDetailsModalContentUserWebsite"
          href="https://google.com "
          target="_blank"
        >
          test.com
        </a>
        <hr
          class="userDetailsModalDivider"
        />
      </div>
      <div
        class="userDetailsModalBodySection"
      >
        <h3
          class="userDetailsModalContentSectionTitle"
        >
          Contact Details
        </h3>
        <p
          class="userDetailsModalContentSectionInfo"
        >
          <a
            href="mailto:jd@jd.jd"
          >
            Email: 
            <b>
              jd@jd.jd
            </b>
             
          </a>
        </p>
        <p
          class="userDetailsModalContentSectionInfo"
        >
          <a
            href="tel:555-5555"
          >
            Phone: 
            <b>
              555-5555
            </b>
          </a>
        </p>
        <p
          class="userDetailsModalContentSectionInfo"
        >
          Username: 
          <b>
            Doe
          </b>
        </p>
        <h3
          class="userDetailsModalContentSectionTitle"
        >
          Company
          : 
          <span
            class="userDetailsModalContentSectionInfo"
          >
            Company
          </span>
        </h3>
        <p
          class="userDetailsModalContentSectionInfo"
        >
          Summary: 
          <b>
            We do it
          </b>
        </p>
        <p
          class="userDetailsModalContentSectionInfo"
        >
          Motto:
           
          <span
            class="userDetailsModalContentSectionInfoQuote"
          >
            '
            Phrase
            '
          </span>
        </p>
        <h3
          class="userDetailsModalContentSectionTitle"
        >
          Address
        </h3>
        <p
          class="userDetailsModalContentSectionInfo"
        >
          <b>
            11 Test Avenue, London
          </b>
        </p>
        <p
          class="userDetailsModalContentSectionInfo"
        >
          <b>
            EC1
          </b>
        </p>
        <p
          class="userDetailsModalContentSectionInfoSmallText"
        >
          test lat, test lng
        </p>
        <button
          class="userDetailsModalCloseButton"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</div>
`;
